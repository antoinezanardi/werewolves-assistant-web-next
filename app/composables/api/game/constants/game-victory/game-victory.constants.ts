const GAME_VICTORY_TYPES = [
  "none",
  "werewolves",
  "villagers",
  "lovers",
  "pied-piper",
  "white-werewolf",
  "angel",
  "prejudiced-manipulator",
] as const;

export { GAME_VICTORY_TYPES };