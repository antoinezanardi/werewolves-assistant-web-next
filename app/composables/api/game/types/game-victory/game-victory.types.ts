import type { TupleToUnion } from "type-fest";

import type { GAME_VICTORY_TYPES } from "~/composables/api/game/constants/game-victory/game-victory.constants";

type GameVictoryType = TupleToUnion<typeof GAME_VICTORY_TYPES>;

export type { GameVictoryType };