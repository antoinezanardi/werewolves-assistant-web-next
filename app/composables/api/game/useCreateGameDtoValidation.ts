import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };