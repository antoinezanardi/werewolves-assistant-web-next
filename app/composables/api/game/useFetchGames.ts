import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  cancelGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  makeGamePlay: (gameId: string, makeGamePlayDto: MakeGamePlayDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  createGameFeedback: (gameId: string, createGameFeedbackDto: CreateGameFeedbackDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch {
      return null;
    }
  }

  async function cancelGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "DELETE" });
    } catch {
      return null;
    }
  }

  async function makeGamePlay(gameId: string, makeGamePlayDto: MakeGamePlayDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/play`, {
        method: "POST",
        body: JSON.stringify(makeGamePlayDto),
      });
    } catch {
      return null;
    }
  }

  async function createGameFeedback(gameId: string, createGameFeedbackDto: CreateGameFeedbackDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/feedback`, {
        method: "POST",
        body: JSON.stringify(createGameFeedbackDto),
      });
    } catch {
      return null;
    }
  }
  return {
    createGame,
    getGame,
    cancelGame,
    makeGamePlay,
    createGameFeedback,
  };
}

export { useFetchGames };