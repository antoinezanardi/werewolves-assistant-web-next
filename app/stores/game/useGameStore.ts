import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };