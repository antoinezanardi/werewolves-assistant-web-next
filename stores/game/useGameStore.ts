import type { AsyncDataRequestStatus } from "#app/composables/asyncData";
import { defineStore } from "pinia";

import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const { getGame: fetchGameFromApi } = useFetchGames();

  const game = ref<Game>(new Game());
  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }
  return {
    game,
    fetchingGameStatus,
    fetchAndSetGame,
  };
});

export { useGameStore };