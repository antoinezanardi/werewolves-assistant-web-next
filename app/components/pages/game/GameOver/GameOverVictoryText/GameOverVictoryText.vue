<template>
  <div
    id="game-over-winners-text"
    class="text-center"
  >
    <div class="flex gap-2 items-center justify-center">
      <NuxtImg
        :alt="$t('components.GameOverVictoryText.trophyImageAlt')"
        :height="svgSize"
        placeholder="/svg/misc/ripples.svg"
        :placeholder-class="`w-[${svgSize}] h-[${svgSize}]`"
        src="svg/misc/trophy.svg"
        :width="svgSize"
      />

      <NuxtImg
        :alt="$t('components.GameOverVictoryText.victoryTypeImageAlt')"
        :height="svgSize"
        placeholder="/svg/misc/ripples.svg"
        :placeholder-class="`w-[${svgSize}] h-[${svgSize}]`"
        :src="victoryTypeTextsAndSvg.svgPath"
        :width="svgSize"
      />
    </div>

    <h1
      id="victory-text"
      class="my-3"
    >
      {{ victoryTypeTextsAndSvg.text }}
    </h1>

    <h4 id="victory-sub-text">
      {{ victoryTypeTextsAndSvg.subText }}
    </h4>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import { VICTORY_TYPES_TEXTS_AND_SVG } from "~/components/pages/game/GameOver/GameOverVictoryText/game-over-victory-text.constants";
import { useAppBreakpoints } from "~/composables/style/useAppBreakpoints";
import { useGameStore } from "~/stores/game/useGameStore";

type TextsAndSvg = {
  text: string;
  subText: string;
  svgPath: string;
};

const { isSmallerThanMdBreakpoint } = useAppBreakpoints();

const svgSize = computed<string>(() => (isSmallerThanMdBreakpoint.value ? "50px" : "125px"));

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const { t } = useI18n();

const isGameWonByLoversAndCupid = computed<boolean>(() => {
  const { victory, options, players } = game.value;
  const alivePlayers = players.filter(player => player.isAlive);
  const teamOfLoveCount = 3;

  return options.roles.cupid.mustWinWithLovers && victory?.type === "lovers" && alivePlayers.length === teamOfLoveCount;
});

const victoryTypeTextsAndSvg = computed<TextsAndSvg>(() => {
  const { victory } = game.value;
  if (isGameWonByLoversAndCupid.value) {
    return {
      text: t("components.GameOverVictoryText.loversAndCupidWin"),
      subText: t("components.GameOverVictoryText.loversAndCupidAreTogetherForever"),
      svgPath: VICTORY_TYPES_TEXTS_AND_SVG.lovers.svgPath,
    };
  }
  if (!victory) {
    return {
      text: "??",
      subText: "??",
      svgPath: "svg/misc/question-mark.svg",
    };
  }
  const { textKey, subTextKey, svgPath } = VICTORY_TYPES_TEXTS_AND_SVG[victory.type];

  return {
    text: t(textKey),
    subText: t(subTextKey),
    svgPath,
  };
});
</script>