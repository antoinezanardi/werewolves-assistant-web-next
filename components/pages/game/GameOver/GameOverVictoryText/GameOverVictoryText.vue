<template>
  <div
    id="game-over-winners-text"
    class="text-center"
  >
    <div class="flex gap-2 items-center justify-center">
      <NuxtImg
        :alt="$t('components.GameOverVictoryText.trophyImageAlt')"
        height="125"
        src="svg/misc/trophy.svg"
        width="125"
      />

      <NuxtImg
        :alt="$t('components.GameOverVictoryText.victoryTypeImageAlt')"
        height="125"
        :src="victoryTypeTextsAndSvg.svgPath"
        width="125"
      />
    </div>

    <h1
      id="victory-text"
      class="my-3"
    >
      {{ victoryTypeTextsAndSvg.text }}
    </h1>

    <h3 id="victory-sub-text">
      {{ victoryTypeTextsAndSvg.subText }}
    </h3>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import { VICTORY_TYPES_TEXTS_AND_SVG } from "~/components/pages/game/GameOver/GameOverVictoryText/game-over-victory-text.constants";
import { useGameStore } from "~/stores/game/useGameStore";

type TextsAndSvg = {
  text: string;
  subText: string;
  svgPath: string;
};

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const { t } = useI18n();

const victoryTypeTextsAndSvg = computed<TextsAndSvg>(() => {
  const { victory } = game.value;
  const unknownVictoryTextsAndSvg: TextsAndSvg = {
    text: "??",
    subText: "??",
    svgPath: "svg/misc/question-mark.svg",
  };
  if (!victory) {
    return unknownVictoryTextsAndSvg;
  }
  const { textKey, subTextKey, svgPath } = VICTORY_TYPES_TEXTS_AND_SVG[victory.type];

  return {
    text: t(textKey),
    subText: t(subTextKey),
    svgPath,
  };
});
</script>