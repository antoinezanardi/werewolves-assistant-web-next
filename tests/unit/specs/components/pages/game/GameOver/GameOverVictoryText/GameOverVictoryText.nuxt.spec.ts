import type { mount } from "@vue/test-utils";

import type { NuxtImg } from "#components";
import GameOverVictoryText from "~/components/pages/game/GameOver/GameOverVictoryText/GameOverVictoryText.vue";
import type { GameVictory } from "~/composables/api/game/types/game-victory/game-victory.class";
import { useGameStore } from "~/stores/game/useGameStore";
import { createFakeGameVictory } from "~/tests/unit/utils/factories/composables/api/game/game-victory/game-victory.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Game Over Victory Text Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameOverVictoryText>>;

  async function mountGameOverVictoryTextComponent(): Promise<ReturnType<typeof mount<typeof GameOverVictoryText>>> {
    return mountSuspendedComponent(GameOverVictoryText);
  }

  beforeEach(async() => {
    wrapper = await mountGameOverVictoryTextComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Trophy Icon", () => {
    it("should render trophy icon when rendered.", () => {
      const trophyIcon = wrapper.findComponent<typeof NuxtImg>("[alt='Trophy icon']");

      expect(trophyIcon.exists()).toBeTruthy();
    });
  });

  describe("Victory Type Icon", () => {
    it.each<{
      victory: GameVictory | undefined;
      expectedIcon: string;
      test: string;
    }>([
      {
        victory: undefined,
        expectedIcon: "svg/misc/question-mark.svg",
        test: "should render question mark icon when victory is undefined.",
      },
      {
        victory: createFakeGameVictory({ type: "werewolves" }),
        expectedIcon: "svg/role/werewolf.svg",
        test: "should render werewolves icon when victory type is werewolves.",
      },
      {
        victory: createFakeGameVictory({ type: "villagers" }),
        expectedIcon: "svg/role/villager.svg",
        test: "should render villagers icon when victory type is villagers.",
      },
      {
        victory: createFakeGameVictory({ type: "lovers" }),
        expectedIcon: "svg/role/lovers.svg",
        test: "should render lovers icon when victory type is lovers.",
      },
      {
        victory: createFakeGameVictory({ type: "angel" }),
        expectedIcon: "svg/role/angel.svg",
        test: "should render angel icon when victory type is angel.",
      },
      {
        victory: createFakeGameVictory({ type: "none" }),
        expectedIcon: "svg/game/player/dead.svg",
        test: "should render none icon when victory type is none.",
      },
      {
        victory: createFakeGameVictory({ type: "pied-piper" }),
        expectedIcon: "svg/role/pied-piper.svg",
        test: "should render pied piper icon when victory type is pied piper.",
      },
      {
        victory: createFakeGameVictory({ type: "prejudiced-manipulator" }),
        expectedIcon: "svg/role/prejudiced-manipulator.svg",
        test: "should render prejudiced manipulator icon when victory type is prejudiced manipulator.",
      },
      {
        victory: createFakeGameVictory({ type: "white-werewolf" }),
        expectedIcon: "svg/role/white-werewolf.svg",
        test: "should render white werewolf icon when victory type is white werewolf.",
      },
    ])("$test", async({ victory, expectedIcon }) => {
      const gameStore = useGameStore();
      gameStore.game.victory = victory;
      await nextTick();
      const victoryTypeIcon = wrapper.findComponent<typeof NuxtImg>("[alt='Victory type icon']");

      expect(victoryTypeIcon.attributes("src")).toBe(expectedIcon);
    });
  });

  describe("Victory Type Text", () => {
    it.each<{
      victory: GameVictory | undefined;
      expectedText: string;
      test: string;
    }>([
      {
        victory: undefined,
        expectedText: "??",
        test: "should render unknown text when victory is undefined.",
      },
      {
        victory: createFakeGameVictory({ type: "werewolves" }),
        expectedText: "components.GameOverVictoryText.werewolvesWin",
        test: "should render werewolves text when victory type is werewolves.",
      },
      {
        victory: createFakeGameVictory({ type: "villagers" }),
        expectedText: "components.GameOverVictoryText.villagersWin",
        test: "should render villagers text when victory type is villagers.",
      },
      {
        victory: createFakeGameVictory({ type: "lovers" }),
        expectedText: "components.GameOverVictoryText.loversWin",
        test: "should render lovers text when victory type is lovers.",
      },
      {
        victory: createFakeGameVictory({ type: "angel" }),
        expectedText: "components.GameOverVictoryText.angelWins",
        test: "should render angel text when victory type is angel.",
      },
      {
        victory: createFakeGameVictory({ type: "none" }),
        expectedText: "components.GameOverVictoryText.nobodyWins",
        test: "should render none text when victory type is none.",
      },
      {
        victory: createFakeGameVictory({ type: "pied-piper" }),
        expectedText: "components.GameOverVictoryText.piedPiperWins",
        test: "should render pied piper text when victory type is pied piper.",
      },
      {
        victory: createFakeGameVictory({ type: "prejudiced-manipulator" }),
        expectedText: "components.GameOverVictoryText.prejudicedManipulatorWins",
        test: "should render prejudiced manipulator text when victory type is prejudiced manipulator.",
      },
      {
        victory: createFakeGameVictory({ type: "white-werewolf" }),
        expectedText: "components.GameOverVictoryText.whiteWerewolfWins",
        test: "should render white werewolf text when victory type is white werewolf.",
      },
    ])("$test", async({ victory, expectedText }) => {
      const gameStore = useGameStore();
      gameStore.game.victory = victory;
      await nextTick();
      const victoryTypeText = wrapper.find<HTMLHeadingElement>("#victory-text");

      expect(victoryTypeText.text()).toBe(expectedText);
    });
  });

  describe("Victory Type SubText", () => {
    it.each<{
      victory: GameVictory | undefined;
      expectedSubText: string;
      test: string;
    }>([
      {
        victory: undefined,
        expectedSubText: "??",
        test: "should render unknown subtext when victory is undefined.",
      },
      {
        victory: createFakeGameVictory({ type: "werewolves" }),
        expectedSubText: "components.GameOverVictoryText.werewolvesHaveEatenALot",
        test: "should render werewolves subtext when victory type is werewolves.",
      },
      {
        victory: createFakeGameVictory({ type: "villagers" }),
        expectedSubText: "components.GameOverVictoryText.villagersHaveSurvivedWerewolves",
        test: "should render villagers subtext when victory type is villagers.",
      },
      {
        victory: createFakeGameVictory({ type: "lovers" }),
        expectedSubText: "components.GameOverVictoryText.loversAreTogetherForever",
        test: "should render lovers subtext when victory type is lovers.",
      },
      {
        victory: createFakeGameVictory({ type: "angel" }),
        expectedSubText: "components.GameOverVictoryText.angelComesBackToParadise",
        test: "should render angel subtext when victory type is angel.",
      },
      {
        victory: createFakeGameVictory({ type: "none" }),
        expectedSubText: "components.GameOverVictoryText.everybodyMurderedEachOther",
        test: "should render none subtext when victory type is none.",
      },
      {
        victory: createFakeGameVictory({ type: "pied-piper" }),
        expectedSubText: "components.GameOverVictoryText.piedPiperHasControl",
        test: "should render pied piper subtext when victory type is pied piper.",
      },
      {
        victory: createFakeGameVictory({ type: "prejudiced-manipulator" }),
        expectedSubText: "components.GameOverVictoryText.prejudicedManipulatorGroupRemainsAlive",
        test: "should render prejudiced manipulator subtext when victory type is prejudiced manipulator.",
      },
      {
        victory: createFakeGameVictory({ type: "white-werewolf" }),
        expectedSubText: "components.GameOverVictoryText.whiteWerewolfWinsAlone",
        test: "should render white werewolf subtext when victory type is white werewolf.",
      },
    ])("$test", async({ victory, expectedSubText }) => {
      const gameStore = useGameStore();
      gameStore.game.victory = victory;
      await nextTick();
      const victoryTypeSubText = wrapper.find<HTMLParagraphElement>("#victory-sub-text");

      expect(victoryTypeSubText.text()).toBe(expectedSubText);
    });
  });
});