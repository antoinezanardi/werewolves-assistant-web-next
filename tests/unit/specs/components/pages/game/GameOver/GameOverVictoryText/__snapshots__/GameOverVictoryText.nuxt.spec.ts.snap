// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Over Victory Text Component > should match snapshot when rendered. 1`] = `
"<div id="game-over-winners-text" class="text-center">
  <div class="flex gap-2 items-center justify-center">
    <nuxt-img-stub alt="Trophy icon" height="125" src="svg/misc/trophy.svg" width="125" preload="false" crossorigin="false" placeholder="/svg/misc/infinite-spinner.svg"></nuxt-img-stub>
    <nuxt-img-stub alt="Victory type icon" height="125" src="svg/misc/question-mark.svg" width="125" preload="false" crossorigin="false" placeholder="/svg/misc/infinite-spinner.svg"></nuxt-img-stub>
  </div>
  <h1 id="victory-text" class="my-3">??</h1>
  <h3 id="victory-sub-text">??</h3>
</div>"
`;
